#                       Analiza i Bazy Danych - Laboratorium 2


Roksana Cieśla, grupa 2





In [1]:
import warnings
warnings.filterwarnings('ignore')

Powyższy fragment kodu służy do wyłączenia ostrzeżeń pojawiających się przy łączeniu się z bazą danych za pomocą pakietów psycopg2 i pandas.

# Łączenie się za pomocą psycopg2 i pandas

In [2]:
import psycopg2 as pg
import pandas as pd

connection = pg.connect(host='pgsql-196447.vipserv.org', port=5432, dbname='wbauer_adb', user='wbauer_adb', password='adb2020')

In [3]:
df = pd.read_sql('select * from city',con=connection)
df

city_id                city  country_id         last_update
0          1  A Corua (La Corua)          87 2006-02-15 09:45:25
1          2                Abha          82 2006-02-15 09:45:25
2          3           Abu Dhabi         101 2006-02-15 09:45:25
3          4                Acua          60 2006-02-15 09:45:25
4          5               Adana          97 2006-02-15 09:45:25
..       ...                 ...         ...                 ...
595      596               Zaria          69 2006-02-15 09:45:25
596      597        Zeleznogorsk          80 2006-02-15 09:45:25
597      598         Zhezqazghan          51 2006-02-15 09:45:25
598      599            Zhoushan          23 2006-02-15 09:45:25
599      600          Ziguinchor          83 2006-02-15 09:45:25

[600 rows x 4 columns]

# 1. Ile kategorii filmów mamy w wypożyczalni?

In [4]:
df = pd.read_sql('select count(distinct name) from category',con=connection)
df

count
0     16

Odpowiedź: Jest 16 kategorii filmów w wypożyczalni.

Poniżej wypisuję zliczone kategorie:

In [5]:
df = pd.read_sql('select distinct name from category',con=connection)
df

name
0        Sports
1      Classics
2           New
3        Family
4        Comedy
5     Animation
6        Travel
7         Music
8        Horror
9         Drama
10       Sci-Fi
11        Games
12  Documentary
13      Foreign
14       Action
15     Children

# 2. Wyświetl listę kategorii w kolejności alfabetycznej.

In [6]:
df = pd.read_sql('select distinct name from category order by name asc',con=connection)
df

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

# 3. Znajdź najstarszy i najmłodszy film do wypożyczenia.

In [7]:
df = pd.read_sql('select title, release_year from film order by release_year asc',con=connection)
df

title  release_year
0      Chamber Italian          2006
1     Grosse Wonderful          2006
2      Airport Pollock          2006
3    Bright Encounters          2006
4     Academy Dinosaur          2006
..                 ...           ...
995     Young Language          2006
996         Youth Kick          2006
997       Zhivago Core          2006
998  Zoolander Fiction          2006
999          Zorro Ark          2006

[1000 rows x 2 columns]

Każdy z filmów do wypożyczenia ma tą sama wartość w kolumnie release_year, zatem nie ma możliwości wybrania nastarszego i najmdłodszego filmu.

# 4. Ile wypożyczeń odbyło się między 2005-07-01 a 2005-08-01?

In [8]:
df = pd.read_sql("select count(rental_date) from rental where rental_date between '2005-07-01' and '2005-08-01'",con=connection)
df

count
0   6709

Odpowiedz: W zadanym okresie odbyło się 6709 wypożyczeń.

# 5. Ile wypożyczeń odbyło się między 2010-01-01 a 2011-02-01?

In [9]:
df = pd.read_sql("select count(rental_date) from rental where rental_date between '2010-01-01' and '2011-02-01'",con=connection)
df

count
0      0

Odpowiedz: W zadanym okresie nie było żadnego wypozyczenia.

# 6. Znajdź największą płatność wypożyczenia.

In [10]:
df = pd.read_sql("select amount from payment order by amount desc limit 1",con=connection)
df

amount
0   11.99

Odpowiedź: Największa płatność wypożyczenia wynosi 11,99.

# 7. Znajdź wszystkich klientów z Polski, Nigerii lub Bangladeszu.

In [11]:
df = pd.read_sql("select * from customer_list where country in ('Poland', 'Nigeria', 'Bangladesh')",con=connection)
df

id              name                         address zip code  \
0   284     Sonia Gregory              1279 Udine Parkway    75860   
1   318       Brian Wyman                 1769 Iwaki Lane    25787   
2   572   Sidney Burleson             1405 Hagonoy Avenue    86587   
3   409    Rodney Moeller          1936 Lapu-Lapu Parkway     7122   
4   332    Stephen Qualls                1838 Tabriz Lane     1195   
5   286       Velma Lucas            1155 Liaocheng Place    22650   
6    80      Marilyn Ross                1888 Kabul Drive    20936   
7   198      Elsie Kelley                 955 Bamenda Way     1545   
8   103   Gladys Hamilton                 1177 Jelets Way     3305   
9   329    Frank Waggoner             1816 Bydgoszcz Loop    64308   
10  128   Marjorie Tucker        1050 Garden Grove Avenue     4999   
11   18      Carol Garcia                320 Brest Avenue    43331   
12  270       Leah Curtis                  906 Goinia Way    83565   
13  501       Ruben Geary          414 Mandaluyong Street    16370   
14  277      Olga Jimenez                556 Baybay Manor    55802   
15  156   Bertha Ferguson               1367 Yantai Manor    21294   
16  280    Tracey Barrett  1006 Santa Brbara dOeste Manor    36229   
17  250         Jo Fowler            786 Stara Zagora Way    98332   
18  571  Johnnie Chisholm      1501 Pangkal Pinang Avenue      943   
19  562     Wallace Slone                 1342 Sharja Way    93655   
20   21    Michelle Clark  262 A Corua (La Corua) Parkway    34418   
21  380   Russell Brinson     1642 Charlotte Amalie Drive    75442   
22  558  Jimmie Eggleston            505 Madiun Boulevard    97271   
23  232    Constance Reid               885 Yingkou Manor    31390   

           phone             city     country   notes  sid  
0   195003555232       Benin City     Nigeria  active    1  
1   556100547674        Bydgoszcz      Poland  active    1  
2   867287719310      Czestochowa      Poland  active    1  
3   653436985797        Deba Habe     Nigeria  active    2  
4    38988715447            Dhaka  Bangladesh  active    1  
5   558236142492     Effon-Alaiye     Nigeria  active    1  
6   701457319790              Ife     Nigeria  active    1  
7   768481779568           Ikerre     Nigeria  active    2  
8   484292626944           Ilorin     Nigeria  active    1  
9   965273813662         Jamalpur  Bangladesh  active    2  
10  973047364353  Jastrzebie-Zdrj      Poland  active    1  
11  747791594069           Kaduna     Nigeria  active    2  
12  701767622697           Kalisz      Poland  active    1  
13   52709222667           Lublin      Poland  active    1  
14  363982224739        Ogbomosho     Nigeria  active    2  
15  889538496300             Ondo     Nigeria  active    1  
16   85059738746              Owo     Nigeria  active    2  
17  716256596301              Oyo     Nigeria  active    2  
18  770864062795            Plock      Poland  active    2  
19  946114054231           Sokoto     Nigeria  active    1  
20  892775750063          Tangail  Bangladesh  active    1  
21  821476736117            Tychy      Poland  active    1  
22  970638808606          Wroclaw      Poland  active    1  
23  588964509072            Zaria     Nigeria  active    2

# 8. Gdzie mieszkają członkowie personelu?

In [12]:
df = pd.read_sql('select distinct country from staff_list',con=connection)
df

country
0  Australia
1     Canada

Odpowiedź: Członkowie personelu mieszkają w Australii i Kanadzie.

# 9. Ilu pracowników mieszka w Argentynie lub Hiszpanii?

In [13]:
df = pd.read_sql("select count(country) from staff_list where country in ('Arentina', 'Spain')",con=connection)
df

count
0      0

Odpowiedź: Żaden z członków personelu nie mieszka w Argentynie ani w Hiszpanii.

# 10. Jakie kategorie filmów zostały wypożyczone przez klientów?

In [14]:
df = pd.read_sql("select distinct category.name from rental inner join inventory on rental.inventory_id = inventory.inventory_id\n" 
                 "inner join film on film.film_id = inventory.film_id\n" 
                 "inner join film_category on film_category.film_id = film.film_id\n" 
                 "inner join category on film_category.category_id = category.category_id",con=connection)
df

name
0        Sports
1      Classics
2           New
3        Family
4        Comedy
5     Animation
6        Travel
7         Music
8         Drama
9        Horror
10       Sci-Fi
11        Games
12  Documentary
13      Foreign
14       Action
15     Children

# 11. Znajdź wszystkie kategorie filmów wypożyczonych w Ameryce.

In [15]:
df = pd.read_sql("select distinct category.name from rental inner join inventory on rental.inventory_id = inventory.inventory_id\n" 
                 "inner join film on film.film_id = inventory.film_id\n" 
                 "inner join film_category on film_category.film_id = film.film_id\n" 
                 "inner join category on film_category.category_id = category.category_id\n" 
                 "inner join customer on customer.customer_id = rental.customer_id\n" 
                 "inner join address on address.address_id = customer.address_id\n" 
                 "inner join city on city.city_id = address.city_id\n" 
                 "inner join country on country.country_id = city.country_id where country.country='United States'",con=connection)
df

name
0        Action
1     Animation
2      Children
3      Classics
4        Comedy
5   Documentary
6         Drama
7        Family
8       Foreign
9         Games
10       Horror
11        Music
12          New
13       Sci-Fi
14       Sports
15       Travel

# 12. Znajdź wszystkie tytuły filmów, w których grał: Olympia Pfeiffer lub Julia Zellweger lub Ellen Presle

In [22]:
df = pd.read_sql("select title, actor.first_name, actor.last_name from film inner join film_actor on film.film_id = film_actor.film_id\n" 
                 "inner join actor on film_actor.actor_id = actor.actor_id\n"
                 "where (actor.first_name='Olympia' and actor.last_name='Pfeiffer') or (actor.first_name='Julia' and actor.last_name='Zellweger')\n" 
                 "or (actor.first_name='Ellen' and actor.last_name='Presley')",con=connection)
df

title first_name  last_name
0     Bilko Anonymous      Ellen    Presley
1   Caribbean Liberty      Ellen    Presley
2    Casper Dragonfly      Ellen    Presley
3    Empire Malkovich      Ellen    Presley
4       Floats Garden      Ellen    Presley
..                ...        ...        ...
64       Outlaw Hanky      Julia  Zellweger
65    Panky Submarine      Julia  Zellweger
66   Rider Caddyshack      Julia  Zellweger
67          Won Dares      Julia  Zellweger
68      Wyoming Storm      Julia  Zellweger

[69 rows x 3 columns]